distributed/tests/test_semaphore.py

```diff
@@ -0,0 +1,156 @@
+from time import sleep
+
+from distributed import Client, Semaphore, get_client
+from distributed.metrics import time
+from distributed.utils_test import cluster, gen_cluster  # noqa F401
+
+import dask
+from dask.distributed import Client
+
+
+@gen_cluster(client=True)
+async def test_semaphore(c, s, a, b):
+    semaphore = await Semaphore(max_leases=2, name="exasol_db")
+
+    result = await semaphore.acquire()  # allowed_leases: 2 - 1 -> 1
+    assert result is True
+
+    second = await semaphore.acquire()  # allowed_leases: 1 - 1 -> 0
+    assert second is True
+    start = time()
+    result = await semaphore.acquire(timeout=0.025)  # allowed_leases: 0 -> False
+    stop = time()
+    assert stop - start < 0.2
+    assert result is False
+
+
+@gen_cluster(client=True)
+async def test_serializable(c, s, a, b):
+    sem = await Semaphore(max_leases=2, name="x")
+    res = await sem.acquire()
+    assert len(s.extensions["semaphores"].leases["x"]) == 1
+    assert res
+    sem2 = pickle.loads(pickle.dumps(sem))
+    assert sem2.name == sem.name
+    assert sem2.client.scheduler.address == sem.client.scheduler.address
+
+    # actual leases didn't change
+    assert len(s.extensions["semaphores"].leases["x"]) == 1
+
+    res = await sem2.acquire()
+    assert res
+
+    # Ensure that both objects access the same semaphore
+    res = await sem.acquire(timeout=0.025)
+
+    assert not res
+    res = await sem2.acquire(timeout=0.025)
+
+    assert not res
+
+
+@gen_cluster(client=True)
+async def test_release_simple(c, s, a, b):
+    def f(x, semaphore):
+        with semaphore:
+            assert semaphore.name == "x"
+            return x + 1
+
+    sem = await Semaphore(max_leases=2, name="x")
+    futures = c.map(f, list(range(10)), semaphore=sem)
+    await c.gather(futures)
+
+
+@gen_cluster(client=True)
+async def test_acquires_with_timeout(c, s, a, b):
+    sem = await Semaphore(1, "x")
+    assert await sem.acquire(timeout=0.025)
+    assert not await sem.acquire(timeout=0.025)
+    await sem.release()
+    assert await sem.acquire(timeout=0.025)
+    await sem.release()
+
+
+def test_timeout_sync():
+    with cluster() as (scheduler, workers):
+        with Client(scheduler["address"]):
+            s = Semaphore(name="x")
+            # Using the context manager already acquires a lease, so the line below won't be able to acquire another one
+            with s:
+                assert s.acquire(timeout=0.025) is False
+
+
+def test_lock_name_only():
+    with cluster() as (scheduler, workers):
+        with Client(scheduler["address"]) as client:
+
+            def f(x):
+                with Semaphore(name="x"):
+                    client = get_client()
+                    assert client.get_metadata("locked") is False
+                    client.set_metadata("locked", True)
+                    sleep(0.01)
+                    assert client.get_metadata("locked") is True
+                    client.set_metadata("locked", False)
```

Comment:
Did you want to run this function many times concurrently?

Reply:
Yes, somewhere a line is lost where `f` is called using, e.g. a `client.map`